sqllogictest: CREATE INDEX on CREATE TABLE

mjibson opened the pull request:
Add a crate that can produce SQL mutations based on some input. Because
it only has a single use at the moment, don't bother with a struct or
configuration yet. Currently it does a single mutation: for each CREATE
TABLE, add a CREATE DEFAULT INDEX on it.

Teach sqllogictest to run this after successful `statement` directives.

On my machine in release mode, the time of running
`sqlite/test/index/orderby/10/slt_good_3.test` went from 3.5 minutes to
36 seconds.

### Motivation

  * This PR fixes a recognized bug. #13786

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-protobuf` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - n/a

mjibson commented:
Some explain plans changed here. Options:

1. rewrite the changed plans
2. add some syntax to opt out of this: `statement no-mutate`. Would be a bit confusing because this is need at the `CREATE`, not the `SELECT`.
3. add a top level file directive that disables it for a whole file

benesch commented:
> rewrite the changed plans

I'd just do this! They're all our own internal tests, right? We can always add (3) as a feature if someone wants to write an SLT that tests an unindexed table, but pretty much all the tests predate unindexed tables, so they actually _want_ to be testing indexed tables.

benesch commented:
Do you have a vision for `mutate-test` evolving into more than it currently does? Maybe I'm not being imaginative enough, but as it is I'd probably just make it a function inside the SLT crate rather than a general-purpose thing with its own crate.

mjibson commented:
> Do you have a vision for `mutate-test` evolving into more than it currently does?

Not really. Testdrive might want it? That was the only reason I made it separate. I don't really care how it's arranged.

mjibson commented:
Made mutate its own function in sqllogictest that we can do something with later if wanted. Added some docs for it so people aren't horribly confused. Rewrote all the plans.

benesch commented on doc/developer/sqllogictest.md:
```diff
@@ -73,6 +73,12 @@ CREATE TABLE bar (
+
+All `statement` directives that contain `CREATE TABLE` inject a `CREATE DEFAULT INDEX ON <t>`.
+This is done to speed up the execution of tests.
+If you need this to not happen, you can run an explicit `DROP INDEX`, or we can design some other way to opt out.
```
Love it.

philip-stoev reviewed (changes requested):
May I suggest a solution that will not add CREATE DEFAULT INDEX to any of our in-house tests so that:
- they will not need to be rewritten
- there will be no merge conflicts with any PRs by the optimizer team currently in flight
- there will be limited possibility for a confusion as to what is being tested in those SLTs going forward and why the EXPLAINS are what they are.

Namely:
- make this rewrite a sqllogictest command-line option
- in `ci/slt/slt.sh` , have a third invocation of sqllogictest that runs only `test/sqllogictest/sqlite/test` with the new option, leaving all the other SLTs to run as they currently do.

mjibson commented:
Changed to be a slt cli option that is invoked for the slow sqllite test directory.
